bindings/java/src/api/com/apple/foundationdb/AbstractFDBException.java

```diff
@@ -0,0 +1,79 @@
+package com.apple.foundationdb;
+
+import com.apple.foundationdb.async.CloneableException;
+
+/**
+ * An Error from FoundationDB.  Each {@code AbstractFDBException} sets
+ *  the {@code message} of the underlying Java {@link Exception}. FDB exceptions expose
+ *  a number of functions including, for example, {@link #isRetryable()} that
+ *  evaluate predicates on the internal FDB error. Most clients should use those methods
+ *  in order to implement special handling for certain errors if their application
+ *  requires it.
+ *
+ * <p>
+ * Errors in FDB should generally be retried if they match the {@link #isRetryable()}
+ *  predicate. In addition, as with any distributed system, certain classes of errors
+ *  may fail in such a way that it is unclear whether the transaction succeeded (they
+ *  {@link #isMaybeCommitted() may be committed} or not). To handle these cases, clients
+ *  are generally advised to make their database operations idempotent and to place
+ *  their operations within retry loops. The FDB Java API provides some default retry loops
+ *  within the {@link Database} interface. See the discussion within the documentation of
+ *  {@link Database#runAsync(Function) Database.runAsync()} for more details.
+ *
+ * @see com.apple.foundationdb.Transaction#onError(Throwable) Transaction.onError()
+ * @see com.apple.foundationdb.Database#runAsync(Function) Database.runAsync()
+ */
+public abstract class AbstractFDBException extends RuntimeException implements CloneableException {
```

Comment:
Feels a bit awkward to make `AbstractFDBException` the abstract version of this class, which I think means that we could technically break someone's code if they had a `catch (FDBException ...)` clause on a `try` block. The alternative, I suppose, is to make `FDBException` `abstract` and introduce a new `FDBExceptionImpl` type or something. (That would, of course, be a breaking change as the constructor on `FDBException` is public.) The alternative alternative might be to remove `FDBException` from the API package at all, and have it live only in the impl package, which I think would mean removing some `@throws` documentation but not much else?

Reply:
It shouldn't break someone's code if they were catching `FDBException` and we are still throwing FDBException in the impl package (that was the intent). Making FDBException abstract would have the benefit of existing code catching FDBException would continue to work if they, say, migrate to the gRPC client (since there we would have to throw a RemoteFDBException and that `FDBException`, still being in the impl package, would not catch properly).

Reply:
Okay, you're trying to be nice to lionrock :) That's fair.